Tutorials for Pod Security Admission

PushkarJ opened the pull request:
Creates a two part tutorial for Pod Security Admission with KinD:
- Applying Pod Security Standards at Cluster level
- Applying Pod Security Standards at Namespace level

/kind documentation

# Notes for reviewers:
- `kind` node image for v1.23 is not yet available here: https://hub.docker.com/r/kindest/node/tags
- So the steps here are documented by building a kind Node image from source
- We could hold this PR until v1.23 tag lands for kind or I can add a step / link on how to build Node image from source
- Once v1.23 lands, can update the tutorial to reflect the correct version instead of `latest` tag

(open to feedback on other ways to tackle this of course :) )

Initial slack discussion: https://kubernetes.slack.com/archives/C1J0BPD2M/p1636152420159200

PushkarJ commented:
/cc @reylejano @shannonxtreme @savitharaghunathan

PushkarJ commented:
/sig security

reylejano commented:
Previews of updated pages:
https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/
https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/pod-security/
https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/pod-security/cluster-level-pss/
https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/pod-security/ns-level-pss/

shannonxtreme commented on content/en/docs/tutorials/_index.md:
```diff
@@ -57,6 +57,11 @@ Before walking through each tutorial, you may want to bookmark the
 
 * [Using Source IP](/docs/tutorials/services/source-ip/)
 
+## Pod Security
```
Rename to Security so it can hold future security related tutorials too.

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+weight: 10
+---
+
+{{% alert title="Disclaimer" %}}
```
Use a note, not a disclaimer. {{\<note\>}}

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+{{% alert title="Disclaimer" %}}
+This tutorial applies only for new clusters.
+{{% /alert %}}
+
```
Start with an overview. This is a one-three sentences that explain what the tutorial teaches you to do.  Something like "This tutorial shows you how to use Pod Security Admission to apply Pod Security Standards to new Kubernetes clusters." (this would remove the need for the Note at the start too)

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+This tutorial applies only for new clusters.
+{{% /alert %}}
+
+# Pre-requisites
```
```suggestion
## {{% heading "prerequisites" %}}
```

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+
+Install the following on your workstation:
+
+- [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
```
Are these the only two things needed to fully complete this tutorial? Nothing else?

PushkarJ replied:
I think so. I have asked one of our newer members to test run this tutorial to confirm!!

PushkarJ replied:
Closing for now. This list also checks out with blog's pre-requisite: https://github.com/kubernetes/website/pull/30502/files#r752655898

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+
+- [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
+- [kubectl](https://kubernetes.io/docs/tasks/tools/)
+
```
```suggestion
## {{% heading "objectives" %}}
```

Include a bulleted list of the objectives. What are the expected outcomes of this tutorial? 

For example: 

* Create a new cluster with Pod Security Standards applied
* Use `--dry-run` to experiment with cluster behavior when standards are applied
* Understand and use admission modes with Pod Security Standards

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+- [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
+- [kubectl](https://kubernetes.io/docs/tasks/tools/)
+
+# Finding the right Pod Security Standard
```
```suggestion

<!-- lessoncontent --> 

## Finding the right Pod Security Standard
```

shannonxtreme replied:
Suggest using: "Choose the Pod Security Standards to apply"

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+
+# Finding the right Pod Security Standard
+
+As of
+v1.23, [Pod Security Admission](/docs/concepts/security/pod-security-admission/)
+allows usage of
+built-in [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
+that can be set in three different modes: `enforce`, `audit`, `warn`.
```
This can go into the overview at the top

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+built-in [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
+that can be set in three different modes: `enforce`, `audit`, `warn`.
+
+Following steps allows gathering information about which Pod Security Standards
+will be most appropriate to apply:
```
```suggestion
Before you apply standards, you should experiment with how your cluster reacts to applying those standards. You can do this by creating a new cluster and using the `--dry-run` flag. 
```

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+for any namespaces. However, `baseline` and `restricted` standards both have
+more warnings specifically on `kube-system` namespace.
+
+# Setting modes, versions and standards
```
```suggestion
## Create and consume a configuration file
```

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
````diff
@@ -0,0 +1,271 @@
+EOF
+```
+
+# Configure the API server
````
This will be a step under the previous heading (so you create the file and configure the API server)

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
```diff
@@ -0,0 +1,271 @@
+directory is added as a Shared Directory under
+`Preferences -> Resources -> File Sharing` (as seen on Docker Desktop for Mac)
+
+# Create cluster with Pod Security Standards
```
```suggestion
## Create a cluster with Pod Security Standards
```

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
````diff
@@ -0,0 +1,271 @@
+To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
+```
+
+# Create pod
````
```suggestion
## Verify the Pod Security Standards configuration 
```

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
````diff
@@ -0,0 +1,271 @@
+pod/nginx created
+```
+
+# Bringing it all together
````
I think you should move this to "What's Next" and say "Run a gist to perform all the preceding steps at once"

shannonxtreme commented on content/en/docs/tutorials/pod-security/cluster-level-pss.md:
````diff
@@ -0,0 +1,271 @@
+Warning: would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" m ...
+pod/nginx created
+```
+
````
\#\# {{% heading "whatsnext" %}}

shannonxtreme replied:
Before What's Next, add \#\# {{% heading "cleanup" %}} if applicable, to show how to remove this and restore to the original state of the cluster.

PushkarJ replied:
Added a clean up section to both tutorials

shannonxtreme commented on content/en/docs/tutorials/pod-security/ns-level-pss.md:
```diff
@@ -0,0 +1,122 @@
+---
+title: Applying Pod Security Standards at Namespace level
```
Are there any major differences between NS level and cluster level? If not, we can simplify this and add the content to the other tutorial. We can say you can apply at cluster or namespace level (and in the relevant commands, tell the user what they'd change to apply to a namespace). If the steps are too different we can even have two sections in that page: Applying at the cluster level and applying at the NS level. 

Let me know what you think.

PushkarJ replied:
Yes IMO, both are pretty different and can be done without following the other. IIRC, except installing KinD and kubectl, most steps are different, but the final outcome might look same. 

Perhaps I can modify the tutorial for NS level and cluster level to show how PSS is applied to all namespaces and for namespace level it can applied one by one. Maybe that will make the difference clearer since it seems the outcome of both the tutorials is identical as they are written today!!

shannonxtreme replied:
That's a good idea, following a common use case for each would be nice. You could explain why you'd apply for one NS at a time and show them how.

PushkarJ replied:
done.

shannonxtreme commented on content/en/docs/tutorials/pod-security/_index.md:
```diff
@@ -0,0 +1,5 @@
+---
+title: "Pod Security"
```
Rename to security

shannonxtreme reviewed (changes requested):
Structural review. I'll do a content review after we have hashed out some of the structural changes :) 

Thank you for this PJ!!

PushkarJ commented:
/hold 

Until merge timelines are aligned as per this comment: https://github.com/kubernetes/website/pull/30502#issuecomment-973301421

PushkarJ commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+This tutorial applies only for new clusters.
+{{% /alert %}}
+
+Since v1.23, Pod Security admission (PSA) is enabled by default as it [graduated
+to beta](blog/_posts/2021-12-15-pod-security-admission-beta.md). In this 
+tutorial, we will enforce `baseline` Pod Security Standard, at a cluster level.
```
@jimangel / @lachie83 can you take a look at the call out for the blog post here and add your comments / suggestions?

shannonxtreme replied:
```suggestion
Since v1.23, Pod Security Admission (PSA) is enabled by default as it [graduated
to beta](blog/_posts/2021-12-15-pod-security-admission-beta.md). Pod Security Admission
is an admission controller that applies Pod Security Standards when pods are created.

This tutorial shows you how to enforce the `baseline` Pod Security Standard at the cluster level.
```

shannonxtreme replied:
One more sentence about the difference between cluster and namespace level. Is it that all namespaces get the same treatment? Link from here to the Namespace tutorial so users can immediately jump to it too.

PushkarJ replied:
done

PushkarJ commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,131 @@
+   `restricted` Pod Security Standard also in `warn` and `audit` mode.
+4. Create a new pod with the following pod security standards applied
+
+> Shoutout to Lachlan Evenson who wrote a
+> [blog](https://medium.com/@LachlanEvenson/hands-on-with-kubernetes-pod-security-admission-b6cac495cd11)
+> that covered similar content with older settings when pod security admission
+> was in alpha state (v1.22)
```
@lachie83 let me know if you would rather keep this or remove it in favor of the call out for official v1.23 blog post? cc @jimangel

jimangel replied:
I like the idea of both sources pointing at each other. If we're planning on this landing on `main` post-release, we might consider throwing a `hold` label on this so it doesn't get merged early.

PushkarJ replied:
@jimangel Yes this is on `/hold` already based on @sftim's suggestion here: https://github.com/kubernetes/website/pull/30502#issuecomment-973931518

The blog links are added here: https://github.com/kubernetes/website/pull/30422#discussion_r752808164 and tutorial links are requested here: https://github.com/kubernetes/website/pull/30502#discussion_r752667699

I can remove the medium blog link if the above cross linking works.

PushkarJ commented:
/sig auth

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+---
+title: Applying Pod Security Standards at Cluster level
```
```suggestion
title: Applying Pod Security Standards at the cluster level
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+This tutorial applies only for new clusters.
+{{% /alert %}}
+
+Since v1.23, Pod Security admission (PSA) is enabled by default as it [graduated
+to beta](blog/_posts/2021-12-15-pod-security-admission-beta.md). In this 
```
```suggestion
Pod Security admission (PSA) is enabled by default in v1.23 and later, as it [graduated
to beta](/blog/2021/12/15/pod-security-admission-beta/). In this 
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+- [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
+- [kubectl](https://kubernetes.io/docs/tasks/tools/)
+
+# Finding the right Pod Security Standard
```
```suggestion
## Choose the right Pod Security Standard to apply
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+# Finding the right Pod Security Standard
+
+As of
+v1.23, [Pod Security Admission](/docs/concepts/security/pod-security-admission/)
+allows usage of
+built-in [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
+that can be set in three different modes: `enforce`, `audit`, `warn`.
```
```suggestion
[Pod Security Admission](/docs/concepts/security/pod-security-admission/)
lets you apply built-in [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
with the following modes: `enforce`, `audit`, and `warn`.
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+built-in [Pod Security Standards](/docs/concepts/security/pod-security-standards/)
+that can be set in three different modes: `enforce`, `audit`, `warn`.
+
+Following steps allows gathering information about which Pod Security Standards
+will be most appropriate to apply:
```
```suggestion
To gather information that helps you to choose the Pod Security Standards
that are most appropriate for your configuration, do the following:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+Following steps allows gathering information about which Pod Security Standards
+will be most appropriate to apply:
+
+1. Create a cluster without any pod security standard applied:
```
```suggestion
1. Create a cluster with no Pod Security Standards applied:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+
+1. Create a cluster without any pod security standard applied:
+
+    ```
+    $ kind create cluster --name psa-wo-cluster-pss --image kindest/node:latest
````
```suggestion
    ```shell
    kind create cluster --name psa-wo-cluster-pss --image kindest/node:latest
    ```
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+
+    ```
+    $ kind create cluster --name psa-wo-cluster-pss --image kindest/node:latest
+    Creating cluster "psa-wo-cluster-pss" ...
````
```suggestion
    The output is similar to this:
    ```
    Creating cluster "psa-wo-cluster-pss" ...
```

Basically, separate output from the command to run.

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+    
+    ```
+
+2. Switch kubectl context to point it to this new cluster
````
```suggestion
1. Set the kubectl context to the new cluster:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+    ```
+
+2. Switch kubectl context to point it to this new cluster
+    ```
+    $ kubectl cluster-info --context kind-psa-wo-cluster-pss
````
```suggestion
    ```shell
    kubectl cluster-info --context kind-psa-wo-cluster-pss
    ```
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+2. Switch kubectl context to point it to this new cluster
+    ```
+    $ kubectl cluster-info --context kind-psa-wo-cluster-pss
+    Kubernetes control plane is running at https://127.0.0.1:61350
````
```suggestion
The output is similar to this:

    ```
     Kubernetes control plane is running at https://127.0.0.1:61350

```

Separate output from commands

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+    To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
+    ```
+
+3. Let's look at existing namespaces in the cluster:
````
```suggestion
1. Get a list of namespaces in the cluster:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+
+3. Let's look at existing namespaces in the cluster:
+
+    ```
````
Separate output from command (follow the same style as previous comments)

PushkarJ replied:
done

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+    local-path-storage   Active   9m26s
+    ```
+
+4. Use `--dry-run=server` to understand what happens when different pod security
+   standard are applied:
````
```suggestion
1. Use `--dry-run=server` to understand what happens when different Pod Security Standards
    are applied:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+4. Use `--dry-run=server` to understand what happens when different pod security
+   standard are applied:
+
+    ```
````
Separate output from commands

PushkarJ replied:
done

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+    namespace/local-path-storage labeled
+    
+    
+    $ kubectl label --dry-run=server --overwrite ns --all \
```
Separate output from commands

PushkarJ replied:
done

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+    namespace/local-path-storage labeled
+    
+    
+    $ kubectl label --dry-run=server --overwrite ns --all \
```
Separate output from commands

PushkarJ replied:
done

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+    namespace/local-path-storage labeled
+    ```
+
+It seems like `privileged` Pod Security Standard when applied, shows no warnings
````
```suggestion
From the previous output, you'll notice that applying the `privileged` Pod Security Standard shows no warnings
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+It seems like `privileged` Pod Security Standard when applied, shows no warnings
+for any namespaces. However, `baseline` and `restricted` standards both have
+more warnings specifically on `kube-system` namespace.
```
```suggestion
warnings, specifically in the `kube-system` namespace.
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+for any namespaces. However, `baseline` and `restricted` standards both have
+more warnings specifically on `kube-system` namespace.
+
+# Setting modes, versions and standards
```
```suggestion
## Set modes, versions and standards
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+# Setting modes, versions and standards
+
+For the purposes of this tutorial, `baseline` standard will be `enforced` as per
+`latest` version whereas `restricted` standard will be applied in
+`warn` and `audit` mode. Baseline Pod Security Standard provides a convenient
+middle ground that allows keeping the exemption list short and prevents known
+privilege escalations.
```
```suggestion
In this tutorial, you apply the following Pod Security Standards to the `latest` version:

  * `baseline` standard in `enforce` mode.
  * `restricted` standard in `warn` and `audit` mode.

The `baseline` Pod Security Standard provides a convenient
middle ground that allows keeping the exemption list short and prevents known
privilege escalations.
```

PS I'm not super clear on what `latest` version means. Latest version of what?

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+middle ground that allows keeping the exemption list short and prevents known
+privilege escalations.
+
+However, to prevent pods from failing in `kube-system`, this namespace will be
+exempt from the cluster level Pod Security Standards applied.
```
```suggestion
Additionally, to prevent pods from failing in `kube-system`, you'll exempt the namespace
from having Pod Security Standards applied.
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+However, to prevent pods from failing in `kube-system`, this namespace will be
+exempt from the cluster level Pod Security Standards applied.
+
+_Notes_:
+
+1. Based on risk posture applied to a cluster, a more strict Pod Security
+   Standard like `restricted` could be a better choice.
+2. This exemption for `kube-system` namespace, allows pods to run as
+   `privileged` in this namespace. So, it is recommended to apply strict RBAC
+   policies that limit access to `kube-system` following least privilege
+   principle.
```
```suggestion
When you implement Pod Security Admission in your own environment, consider the
following:

1. Based on the risk posture applied to a cluster, a stricter Pod Security
   Standard like `restricted` might be a better choice.
1. Exempting the `kube-system` namespace allows pods to run as
   `privileged` in this namespace. We recommend that you apply strict RBAC
   policies that limit access to `kube-system`, following the principle of least
   privilege.
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+   policies that limit access to `kube-system` following least privilege
+   principle.
+
+Let's create a configuration file that can be consumed by built-in Pod Security
```
```suggestion
1. Create a configuration file that can be consumed by the Pod Security
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+Admission Controller to implement these Pod Security Standards:
+
+```
+$ mkdir -p /tmp/pss
+$ cat <<EOF > /tmp/pss/cluster-level-pss.yaml 
````
```suggestion
mkdir -p /tmp/pss
cat <<EOF > /tmp/pss/cluster-level-pss.yaml 
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+EOF
+```
+
+# Configure the API server
````
```suggestion
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+# Configure the API server
+
+Let's configure API server to consume this file during cluster creation:
```
```suggestion
1. Configure the API server to consume this file during cluster creation:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+Let's configure API server to consume this file during cluster creation:
+
+```
+$ cat <<EOF > /tmp/pss/cluster-config.yaml 
````
```suggestion
cat <<EOF > /tmp/pss/cluster-config.yaml 
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+EOF
+```
+
+_Note_: Please ensure that when using Docker Desktop with KinD, the `/tmp`
+directory is added as a Shared Directory under
+`Preferences -> Resources -> File Sharing` (as seen on Docker Desktop for Mac)
````
```suggestion
{{<note>}}
If you use Docker Desktop with KinD, the `/tmp`
directory is added as a Shared Directory under
**Preferences > Resources > File Sharing** on Mac OS.
{{</note>}}
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+directory is added as a Shared Directory under
+`Preferences -> Resources -> File Sharing` (as seen on Docker Desktop for Mac)
+
+# Create cluster with Pod Security Standards
```
```suggestion
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+# Create cluster with Pod Security Standards
+
+Now we are ready to create a cluster that consumes these pod security standards
+configured with pod security admission controller:
```
```suggestion
1. Create a cluster that uses Pod Security Admission to apply
   these Pod Security Standards:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+configured with pod security admission controller:
+
+```
+$ kind create cluster --name psa-with-cluster-pss --image kindest/node:latest --config /tmp/pss/cluster-config.yaml
````
```suggestion
kind create cluster --name psa-with-cluster-pss --image kindest/node:latest --config /tmp/pss/cluster-config.yaml
```

Separate the output from the command

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
+```
+
+# Create pod
````
```suggestion
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+# Create pod
+
+Creating a pod with minimal configuration in default namespace should now be
+successful with warnings for `restricted` pod security standards:
```
```suggestion
1. Create a Pod with minimal configuration in the default namespace:
```

Would this not have been possible before? We want to at least say what would be different in the original, non-PSA cluster.

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+successful with warnings for `restricted` pod security standards:
+
+```
+$ cat <<EOF > /tmp/pss/nginx-pod.yaml
````
```suggestion
cat <<EOF > /tmp/pss/nginx-pod.yaml
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+EOF
+
+
+$ kubectl apply -f /tmp/pss/nginx-pod.yaml
```
Different command - new step called "Apply the configuration to your cluster:"

PushkarJ replied:
Done

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+
+
+$ kubectl apply -f /tmp/pss/nginx-pod.yaml
+Warning: would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" m ...
```
Separate output from command

PushkarJ replied:
Done

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,280 @@
+Warning: would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "nginx" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "nginx" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "nginx" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "nginx" m ...
+pod/nginx created
+```
+# Clean up
````
```suggestion
## Clean up
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+# Clean up
+
+Run `kind delete cluster -name psa-with-cluster-pss` and 
+`kind delete cluster -name psa-wo-cluster-pss` to delete the clusters we 
```
```suggestion
`kind delete cluster -name psa-wo-cluster-pss` to delete the clusters you 
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+`kind delete cluster -name psa-wo-cluster-pss` to delete the clusters we 
+created.
+
+# Bringing it all together
```
```suggestion
## Bringing it all together
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,280 @@
+4. Set kubectl context to this new cluster
+5. Create a minimal pod yaml file
+6. Apply this file to create a Pod in the new cluster
+
```
Add a ## {{% heading "whatsnext" %}} 

and a link to:

Pod Security Admission (/docs/concepts/security/pod-security-admission/)
Pod Security Standards (/docs/concepts/security/pod-security-standards/)
Applying Pod Security Standards at the namespace level (/docs/tutorials/security/ns-level-pss/)

PushkarJ replied:
done

shannonxtreme reviewed (changes requested):
Went through for style, structure, etc :) This looks awesome so far PJ, thank you for doing this!

PushkarJ commented:
Thank you so much for the thoughtful feedback on the PR @shannonxtreme . All the feedback should be resolved with recent changes. It also refers to the link of blog post from #30502 right at the top of both tutorials (cc @jimangel @lachie83). 

Only pending item I can think of is using `busybox` instead of `nginx` pod container image. I am happy to switch it post publication in Jan (since I am on PTO rest of December), if there are no concerns. 

PR is currently on `/hold` for work on Blog PR to wrap up. Once that PR is merged, we can `/unhold` this assuming no more work or feedback is needed here :) Excited to se this all come together !!!

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+tutorial](/docs/tutorials/security/ns-level-pss).
+
+## {{% heading "prerequisites" %}}
+
```
nit: I'd tell readers that they need either Docker or Podman to practice this.

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,319 @@
+    ```
+
+    {{<note>}}
+    If you use Docker Desktop with KinD, the `/tmp`
+    directory is added as a Shared Directory under
+    **Preferences > Resources > File Sharing** on Mac OS.
````
nit:
```suggestion
    If you use Docker Desktop with KinD on macOS, you can
    add `/tmp` as a Shared Directory under the menu item
    **Preferences > Resources > File Sharing**.
```

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+## {{% heading "whatsnext" %}}
+
+- Run a 
+  [gist](https://gist.github.com/PushkarJ/9f7a0045f4bec31097bdd1e9db0f2f6e) 
```
Please move this from a gist into an example you commit (into `content/en/examples`), or omit the advice. It's OK to link to the example script on GitHub if that's useful to readers.
We definitely prefer not to recommend running 3rd party code from gists we don't control.

PushkarJ replied:
done

sftim commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+## {{% heading "whatsnext" %}}
+
+- Run a
+[gist](https://gist.github.com/PushkarJ/c694bac35c2d100f906861667474afb5)
```
Please move this from a gist into an example you commit (into `content/en/examples`), or omit the advice. It's OK to link to the example script on GitHub if that's useful to readers.
We definitely prefer not to recommend running 3rd party code from gists we don't control.

PushkarJ replied:
Done

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+---
+title: Applying Pod Security Standards at the cluster level
+content_type: tutorial
+weight: 10
```
nit: both articles have the same weight, is that intended?

PushkarJ replied:
Yes

sftim commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+one namespace at a time.
+
+# Pre-requisites
+
```
nit: I'd tell readers that they need either Docker or Podman to practice this.

sftim commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+Pod Security admission (PSA) is enabled by default in v1.23 and later, as it [graduated
+to beta](/blog/2021/12/15/pod-security-admission-beta/). Pod Security Admission
+is an admission controller that applies Pod Security Standards when pods are
+created. In this tutorial, we will enforce `baseline` Pod Security Standard, 
```
nit:
```suggestion
created. In this tutorial, you will enforce the `baseline` Pod Security Standard,
```

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+1. Based on the risk posture applied to a cluster, a stricter Pod Security
+   Standard like `restricted` might be a better choice.
+1. Exempting the `kube-system` namespace allows pods to run as
+   `privileged` in this namespace. We recommend that you apply strict RBAC
```
```suggestion
   `privileged` in this namespace. For real world use, the Kubernetes project
   strongly recommends that you apply strict RBAC
```
Maybe mention that equivalent security measures are OK too?

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+---
+title: Applying Pod Security Standards at the cluster level
```
nit (page titles are title case)
```suggestion
title: Apply Pod Security Standards at the Cluster Level
```

sftim commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+---
+title: Applying Pod Security Standards at Namespace level
```
nit (page titles are title case)
```suggestion
title: Apply Pod Security Standards at the Namespace Level
```

sftim reviewed (commented):
Mostly inline feedback. Things I marked as a nit wouldn't block a merge.

overall nit: I'd put these tutorials in the `clusters` directory, and then raise an issue (post-merge) to rename `clusters` to `security`. The AppArmor and seccomp tutorials are just as much about security as the PSA one.

PushkarJ commented:
> Mostly inline feedback. Things I marked as a nit wouldn't block a merge.
> 
> overall nit: I'd put these tutorials in the `clusters` directory, and then raise an issue (post-merge) to rename `clusters` to `security`. The AppArmor and seccomp tutorials are just as much about security as the PSA one.

Thank you @sftim for helping me discover the examples directory under docs :) I did not know it existed before. Removed the references to gists and moved them to examples. Also accepted vast majority of the nits.

Issue for merging clusters tutorials with security is also here: https://github.com/kubernetes/website/issues/30778

PushkarJ commented:
/label tide/merge-method-squash

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+Standard at the cluster level which applies a standard configuration 
+to all namespaces in a cluster.
+
+For applying pod security standards one namespace at a time, please [follow this
+tutorial](/docs/tutorials/security/ns-level-pss).
```
```suggestion
To apply Pod Security Standards to specific namespaces, refer to [Apply Pod Security Standards at the namespace level](/docs/tutorials/security/ns-level-pss).
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,319 @@
+    
+    ```
+
+2. Set the kubectl context to the new cluster:
````
nit, but specifying `1.` for every number in an ordered list (instead of 1 2 3 4 etc) in Markdown reduces future maintenance overhead, since list items can be added/removed without having to adjust the rest of the items in the list.

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+
+## Set modes, versions and standards
+
+In this tutorial, you apply the following Pod Security Standards to the `latest` version:
```
```suggestion
In this section, you apply the following Pod Security Standards to the `latest` version:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+   strongly recommends that you apply strict RBAC
+   policies that limit access to `kube-system`, following the principle of least
+   privilege.
+
```
```suggestion
To implement the preceding standards, do the following:
```

(Introduce these steps and fixes the numbering from continuing)

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,319 @@
+    EOF
+    ```
+
+4. Configure the API server to consume this file during cluster creation:
````
```suggestion
1. Configure the API server to consume this file during cluster creation:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+    **Preferences > Resources > File Sharing**.
+    {{</note>}}
+
+5. Create a cluster that uses Pod Security Admission to apply
```
```suggestion
1. Create a cluster that uses Pod Security Admission to apply
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+    
+    kubectl cluster-info --context kind-psa-with-cluster-pss
+    
+    Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
```
```suggestion
    Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
    ```
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+    
+    Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂
+    
+6. Point kubectl to the cluster 
```
```suggestion
1. Point kubectl to the cluster 
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,319 @@
+6. Point kubectl to the cluster 
+   ```shell
+    kubectl cluster-info --context kind-psa-with-cluster-pss
+    Kubernetes control plane is running at https://127.0.0.1:63855
````
Separate output from the command

PushkarJ replied:
Done. Thanks for picking this one.

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,319 @@
+    To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
+    ```
+
+7. Create a Pod with minimal configuration in the default namespace:
````
```suggestion
1. Create the following Pod specification for a minimal configuration in the default namespace:
```

shannonxtreme commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,319 @@
+            - containerPort: 80
+    EOF
+    ```
+8. Create Pod after pod security is enabled at cluster level:
````
```suggestion
1. Create the Pod in the cluster:
```

shannonxtreme replied:
You don't need to say "after enabling" because that's what these steps did anyway

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+created. In this tutorial, you will enforce the `baseline` Pod Security Standard,
+one namespace at a time.
+
+# Pre-requisites
```
```suggestion
## {{% heading "prerequisites" %}}
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+    Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/
+    ```
+
+2. Set the kubectl context to the new cluster
````
```suggestion
1. Set the kubectl context to the new cluster:
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+- [KinD](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
+- [kubectl](https://kubernetes.io/docs/tasks/tools/)
+
+# Create cluster
```
```suggestion
## Create cluster
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+    To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
+    ```
+
+# Create Namespace
````
```suggestion
## Create a namespace
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+
+# Create Namespace
+
+Create a new namespace `example` for this tutorial:
```
```suggestion
Create a new namespace called `example`:
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+namespace/example created
+```
+
+# Applying one Pod Security Standard
````
I don't see the value in instructions for both one and multiple Standards. Make a single section called "Apply Pod Security Standards" and keep the instructions for applying multiple.

sftim replied:
I'd be OK with adjusting this post-merge

PushkarJ replied:
Merged the two sections together!

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+  pod-security.kubernetes.io/audit-version=latest
+```
+
+# Create Pod
````
```suggestion
## Verify the Pod Security Standards
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+    pod/nginx created
+    ```
+
+As you can see the Pod Security Standards were applied only to `example` 
+namespace. For `default` namespace, pod was created without any warnings. 
+To apply pod security standards to multiple namespaces at once at cluster 
+level, please 
+[follow this tutorial](/docs/tutorials/security/cluster-level-pss).
````
```suggestion
The Pod Security Standards were applied only to the `example` 
namespace. You could create the same Pod in the `default` namespace
with no warnings. 
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+is an admission controller that applies Pod Security Standards when pods are
+created. In this tutorial, you will enforce the `baseline` Pod Security Standard,
+one namespace at a time.
+
```
```suggestion

You can also apply Pod Security Standards to multiple namespaces at once at the cluster 
level. For instructions, refer to [Apply Pod Security Standards at the cluster level](/docs/tutorials/security/cluster-level-pss).
```

shannonxtreme commented on content/en/docs/tutorials/security/ns-level-pss.md:
```diff
@@ -0,0 +1,160 @@
+level, please 
+[follow this tutorial](/docs/tutorials/security/cluster-level-pss).
+
+# Clean up
```
```suggestion
## Clean up
```

shannonxtreme reviewed (changes requested):
Final set of edits and I think we have a decent MVP!

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+This tutorial applies only for new clusters.
+{{% /alert %}}
+
+Pod Security admission (PSA) is enabled by default in v1.23 and later, as it [graduated
+to beta](/blog/2021/12/15/pod-security-admission-beta/). Pod Security Admission
```
https://github.com/kubernetes/website/pull/30502 has the article due for publication on 2021-12-09

```suggestion
Pod Security admission (PSA) is enabled by default in v1.23 and later, as it has
[graduated to beta](/blog/2021/12/09/pod-security-admission-beta/).
Pod Security 
```

sftim commented on content/en/docs/tutorials/security/cluster-level-pss.md:
```diff
@@ -0,0 +1,319 @@
+
+Pod Security admission (PSA) is enabled by default in v1.23 and later, as it [graduated
+to beta](/blog/2021/12/15/pod-security-admission-beta/). Pod Security Admission
+is an admission controller that applies Pod Security Standards when pods are 
```
nit: ```suggestion
is an admission controller that carries out checks against the Kubernetes
[Pod Security Standards](docs/concepts/security/pod-security-standards/) when new pods are
```

sftim reviewed (changes requested):
One actually non-blocking but crucial broken hyperlink, one nit.

sftim commented:
/approve

The _intent_ of this PR is right on target.

jimangel commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,321 @@
+1. Create a cluster with no Pod Security Standards applied:
+
+    ```shell
+    kind create cluster --name psa-wo-cluster-pss --image kindest/node:latest
````
```suggestion
    kind create cluster --name psa-wo-cluster-pss --image kindest/node:v1.23.0
```

jimangel commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,321 @@
+    The output is similar to this:
+    ```
+    Creating cluster "psa-wo-cluster-pss" ...
+    ✓ Ensuring node image (kindest/node:latest) 🖼
````
```suggestion
    ✓ Ensuring node image (kindest/node:v1.23.0) 🖼
```

jimangel commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,321 @@
+   these Pod Security Standards:
+
+   ```shell
+    kind create cluster --name psa-with-cluster-pss --image kindest/node:latest --config /tmp/pss/cluster-config.yaml
````
```suggestion
    kind create cluster --name psa-with-cluster-pss --image kindest/node:v1.23.0 --config /tmp/pss/cluster-config.yaml
```

jimangel commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,321 @@
+   The output is similar to this:
+   ```
+    Creating cluster "psa-with-cluster-pss" ...
+     ✓ Ensuring node image (kindest/node:latest) 🖼 
````
```suggestion
     ✓ Ensuring node image (kindest/node:v1.23.0) 🖼 
```

jimangel commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+1. Create a `KinD` cluster as follows:
+
+    ```shell
+    kind create cluster --name psa-ns-level --image kindest/node:latest
````
```suggestion
    kind create cluster --name psa-ns-level --image kindest/node:v1.23.0
```

jimangel commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,160 @@
+    The output is similar to this:
+    ```
+    Creating cluster "psa-ns-level" ...
+     ✓ Ensuring node image (kindest/node:latest) 🖼 
````
```suggestion
     ✓ Ensuring node image (kindest/node:v1.23.0) 🖼 
```

jimangel commented on content/en/examples/security/kind-with-cluster-level-baseline-pod-security.sh:
```diff
@@ -0,0 +1,72 @@
+    # default None
+    propagation: None
+EOF
+# Until v1.23 is released, kind node image needs to be built from k/k master branch
+# Ref: https://kind.sigs.k8s.io/docs/user/quick-start/#building-images
+kind create cluster --name psa-with-cluster-pss --image kindest/node:latest --config /tmp/pss/cluster-config.yaml
```
```suggestion
kind create cluster --name psa-with-cluster-pss --image kindest/node:v1.23.0 --config /tmp/pss/cluster-config.yaml
```

jimangel commented on content/en/examples/security/kind-with-namespace-level-baseline-pod-security.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
+# Until v1.23 is released, kind node image needs to be built from k/k master branch
+# Ref: https://kind.sigs.k8s.io/docs/user/quick-start/#building-images
+kind create cluster --name psa-ns-level --image kindest/node:latest
```
```suggestion
kind create cluster --name psa-ns-level --image kindest/node:v1.23.0
```

jimangel reviewed (commented):
Update to a valid 1.23 kind version.

PushkarJ commented:
Thank you Jim, Tim and Shannon. Accepted all inline changes from phone app. Will do some final clean up today/tomorrow when I get on Wi-Fi. After that we should be in good shape to publish this together with the blog!

PushkarJ commented:
Alright, just pushed the final changes that takes care of all the pending actionable feedback. 

Now that blog post PR is merged, just need someone to `/hold remove` and `/lgtm` on or before publish date for the blog (Dec 09)

sftim commented:
/check-cla
/easycla

sftim commented:
Hold from https://github.com/kubernetes/website/pull/30422#issuecomment-973654356 should stand until the blog article publishes.

Any time after **16:05 Pacific time on the 8th of December** should be good to go, as the related blog article goes live at **UTC midnight on the 9th**.
It's also OK to PR an update to the publication time of that article if that makes the sync easier.

sftim commented:
This is ready for a technical signoff from SIG Auth. It's already been through quite a few checks, including by me, so what I'm looking for is a final read-through and a formal /lgtm providing no concerns spotted.
In other words, there is no need to go through each step to verify.

Page previews:
- [Apply Pod Security Standards at Cluster level](https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/security/cluster-level-pss/)
- [Apply Pod Security Standards at Namespace level](https://deploy-preview-30422--kubernetes-io-main-staging.netlify.app/docs/tutorials/security/ns-level-pss/)

LGTM for SIG Docs, and https://github.com/kubernetes/website/pull/30422#pullrequestreview-815339185 implies it also looks good to @shannonxtreme

sftim commented:
/remove-label tide/merge-method-squash

Commits are already squashed

sftim commented on content/en/examples/security/kind-with-namespace-level-baseline-pod-security.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
+# Until v1.23 is released, kind node image needs to be built from k/k master branch
+# Ref: https://kind.sigs.k8s.io/docs/user/quick-start/#building-images
```
femtonit: the release has happened (very happy to fix this in a follow-up PR, though)

liggitt commented on content/en/docs/tutorials/security/cluster-level-pss.md:
````diff
@@ -0,0 +1,324 @@
+1. Configure the API server to consume this file during cluster creation:
+
+    ```
+    cat <<EOF > /tmp/pss/cluster-config.yaml 
````
is /tmp a good location for cluster confit files? I'm not sure what the typical permissions there are (or if files in /tmp are subject to automated cleanup/removal)

liggitt commented on content/en/examples/security/kind-with-cluster-level-baseline-pod-security.sh:
```diff
@@ -0,0 +1,70 @@
+#!/bin/sh
+mkdir -p /tmp/pss
+cat <<EOF > /tmp/pss/cluster-level-pss.yaml
```
same question about using /tmp

liggitt commented on content/en/docs/tutorials/security/ns-level-pss.md:
````diff
@@ -0,0 +1,159 @@
+      pod-security.kubernetes.io/warn-version=latest
+    ```
+
+2. Multiple pod security standards can be enabled on any namespace, using labels.
````
it might be worth noting that if this is an existing namespace that already contains workloads, using --dry-run=server is recommended first to determine if the new policy levels will disrupt existing workloads

liggitt commented:
Technical content lgtm, had a non-blocking question about use of tmp and suggestion about calling out use of dry-run

sftim commented:
Thanks @liggitt 

Taking https://github.com/kubernetes/website/pull/30422#issuecomment-988840804 as
/lgtm

We can do a follow-up PR. I agree about taking care about using `/tmp`. There might be a way to improve on this.

shannonxtreme commented:
LGTM too PJ, wonderful work here ❤️

On Wed., Dec. 8, 2021, 18:09 Kubernetes Prow Robot, <
***@***.***> wrote:

> LGTM label has been added.
> Git tree hash: 377841f45aca39355245243a3391e41bda229ba5
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/website/pull/30422#issuecomment-988846119>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHH4EFXXTM47JCWCLNJRBZDUP5RJLANCNFSM5HV7LZTQ>
> .
>

PushkarJ commented:
Blog is now published https://kubernetes.io/blog/2021/12/09/pod-security-admission-beta/

/hold remove

PushkarJ commented:
/hold cancel
